// @kitql/helpers/cjs/colors/stylesNode.js
"use strict";
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var stylesNode_exports = {};
__export(stylesNode_exports, {
  bgBlack: () => bgBlack,
  bgBlackBright: () => bgBlackBright,
  bgBlue: () => bgBlue,
  bgBlueBright: () => bgBlueBright,
  bgCyan: () => bgCyan,
  bgCyanBright: () => bgCyanBright,
  bgGreen: () => bgGreen,
  bgGreenBright: () => bgGreenBright,
  bgMagenta: () => bgMagenta,
  bgMagentaBright: () => bgMagentaBright,
  bgRed: () => bgRed,
  bgRedBright: () => bgRedBright,
  bgWhite: () => bgWhite,
  bgWhiteBright: () => bgWhiteBright,
  bgYellow: () => bgYellow,
  bgYellowBright: () => bgYellowBright,
  black: () => black,
  blackBright: () => blackBright,
  blue: () => blue,
  blueBright: () => blueBright,
  bold: () => bold,
  cyan: () => cyan,
  cyanBright: () => cyanBright,
  dim: () => dim,
  gray: () => gray,
  green: () => green,
  greenBright: () => greenBright,
  hidden: () => hidden,
  inverse: () => inverse,
  italic: () => italic,
  magenta: () => magenta,
  magentaBright: () => magentaBright,
  red: () => red,
  redBright: () => redBright,
  reset: () => reset,
  strikethrough: () => strikethrough,
  underline: () => underline,
  white: () => white,
  whiteBright: () => whiteBright,
  yellow: () => yellow,
  yellowBright: () => yellowBright
});
module.exports = __toCommonJS(stylesNode_exports);
const reset = ["\x1B[0m", "\x1B[0m"];
const bold = ["\x1B[1m", "\x1B[22m"];
const dim = ["\x1B[2m", "\x1B[22m"];
const italic = ["\x1B[3m", "\x1B[23m"];
const underline = ["\x1B[4m", "\x1B[24m"];
const inverse = ["\x1B[7m", "\x1B[27m"];
const hidden = ["\x1B[8m", "\x1B[28m"];
const strikethrough = ["\x1B[9m", "\x1B[29m"];
const black = ["\x1B[30m", "\x1B[39m"];
const red = ["\x1B[31m", "\x1B[39m"];
const green = ["\x1B[32m", "\x1B[39m"];
const yellow = ["\x1B[33m", "\x1B[39m"];
const blue = ["\x1B[34m", "\x1B[39m"];
const magenta = ["\x1B[35m", "\x1B[39m"];
const cyan = ["\x1B[36m", "\x1B[39m"];
const white = ["\x1B[37m", "\x1B[39m"];
const gray = ["\x1B[90m", "\x1B[39m"];
const bgBlack = ["\x1B[40m", "\x1B[49m"];
const bgRed = ["\x1B[41m", "\x1B[49m"];
const bgGreen = ["\x1B[42m", "\x1B[49m"];
const bgYellow = ["\x1B[43m", "\x1B[49m"];
const bgBlue = ["\x1B[44m", "\x1B[49m"];
const bgMagenta = ["\x1B[45m", "\x1B[49m"];
const bgCyan = ["\x1B[46m", "\x1B[49m"];
const bgWhite = ["\x1B[47m", "\x1B[49m"];
const blackBright = ["\x1B[90m", "\x1B[39m"];
const redBright = ["\x1B[91m", "\x1B[39m"];
const greenBright = ["\x1B[92m", "\x1B[39m"];
const yellowBright = ["\x1B[93m", "\x1B[39m"];
const blueBright = ["\x1B[94m", "\x1B[39m"];
const magentaBright = ["\x1B[95m", "\x1B[39m"];
const cyanBright = ["\x1B[96m", "\x1B[39m"];
const whiteBright = ["\x1B[97m", "\x1B[39m"];
const bgBlackBright = ["\x1B[100m", "\x1B[49m"];
const bgRedBright = ["\x1B[101m", "\x1B[49m"];
const bgGreenBright = ["\x1B[102m", "\x1B[49m"];
const bgYellowBright = ["\x1B[103m", "\x1B[49m"];
const bgBlueBright = ["\x1B[104m", "\x1B[49m"];
const bgMagentaBright = ["\x1B[105m", "\x1B[49m"];
const bgCyanBright = ["\x1B[106m", "\x1B[49m"];
const bgWhiteBright = ["\x1B[107m", "\x1B[49m"];
