// @kitql/helpers/cjs/colors/stylesBrowser.js
"use strict";
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var stylesBrowser_exports = {};
__export(stylesBrowser_exports, {
  bgBlack: () => bgBlack,
  bgBlackBright: () => bgBlackBright,
  bgBlue: () => bgBlue,
  bgBlueBright: () => bgBlueBright,
  bgCyan: () => bgCyan,
  bgCyanBright: () => bgCyanBright,
  bgGreen: () => bgGreen,
  bgGreenBright: () => bgGreenBright,
  bgMagenta: () => bgMagenta,
  bgMagentaBright: () => bgMagentaBright,
  bgRed: () => bgRed,
  bgRedBright: () => bgRedBright,
  bgWhite: () => bgWhite,
  bgWhiteBright: () => bgWhiteBright,
  bgYellow: () => bgYellow,
  bgYellowBright: () => bgYellowBright,
  black: () => black,
  blackBright: () => blackBright,
  blue: () => blue,
  blueBright: () => blueBright,
  bold: () => bold,
  cyan: () => cyan,
  cyanBright: () => cyanBright,
  dim: () => dim,
  gray: () => gray,
  green: () => green,
  greenBright: () => greenBright,
  hidden: () => hidden,
  inverse: () => inverse,
  italic: () => italic,
  magenta: () => magenta,
  magentaBright: () => magentaBright,
  red: () => red,
  redBright: () => redBright,
  reset: () => reset,
  strikethrough: () => strikethrough,
  underline: () => underline,
  white: () => white,
  whiteBright: () => whiteBright,
  yellow: () => yellow,
  yellowBright: () => yellowBright
});
module.exports = __toCommonJS(stylesBrowser_exports);
const reset = "";
const bold = "font-weight: bold";
const dim = "opacity: 0.5";
const italic = "font-style: italic";
const underline = "text-decoration: underline";
const inverse = "filter: invert(1)";
const hidden = "visibility: hidden";
const strikethrough = "text-decoration: line-through";
const black = "color: black";
const red = "color: red";
const green = "color: green";
const yellow = "color: yellow";
const blue = "color: blue";
const magenta = "color: #ff00ff";
const cyan = "color: cyan";
const white = "color: white";
const gray = "color: gray";
const bgBlack = "background-color: black";
const bgRed = "background-color: red";
const bgGreen = "background-color: green";
const bgYellow = "background-color: yellow";
const bgBlue = "background-color: blue";
const bgMagenta = "background-color: #ff00ff";
const bgCyan = "background-color: cyan";
const bgWhite = "background-color: white";
const blackBright = "color: #a9a9a9";
const redBright = "color: #ff4500";
const greenBright = "color: #7fff00";
const yellowBright = "color: #ffd700";
const blueBright = "color: #1e90ff";
const magentaBright = "color: #ff69b4";
const cyanBright = "color: #00ffff";
const whiteBright = "color: #ffffff";
const bgBlackBright = "background-color: #a9a9a9";
const bgRedBright = "background-color: #ff4500";
const bgGreenBright = "background-color: #7fff00";
const bgYellowBright = "background-color: #ffd700";
const bgBlueBright = "background-color: #1e90ff";
const bgMagentaBright = "background-color: #ff69b4";
const bgCyanBright = "background-color: #00ffff";
const bgWhiteBright = "background-color: #ffffff";
