// @kitql/helpers/cjs/index.js
"use strict";
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var index_exports = {};
__export(index_exports, {
  Log: () => import_Log.Log,
  bgBlack: () => import_colors.bgBlack,
  bgBlackBright: () => import_colors.bgBlackBright,
  bgBlue: () => import_colors.bgBlue,
  bgBlueBright: () => import_colors.bgBlueBright,
  bgCyan: () => import_colors.bgCyan,
  bgCyanBright: () => import_colors.bgCyanBright,
  bgGreen: () => import_colors.bgGreen,
  bgGreenBright: () => import_colors.bgGreenBright,
  bgMagenta: () => import_colors.bgMagenta,
  bgMagentaBright: () => import_colors.bgMagentaBright,
  bgRed: () => import_colors.bgRed,
  bgRedBright: () => import_colors.bgRedBright,
  bgWhite: () => import_colors.bgWhite,
  bgWhiteBright: () => import_colors.bgWhiteBright,
  bgYellow: () => import_colors.bgYellow,
  bgYellowBright: () => import_colors.bgYellowBright,
  black: () => import_colors.black,
  blackBright: () => import_colors.blackBright,
  blue: () => import_colors.blue,
  blueBright: () => import_colors.blueBright,
  bold: () => import_colors.bold,
  cyan: () => import_colors.cyan,
  cyanBright: () => import_colors.cyanBright,
  dateCompact: () => import_date.dateCompact,
  dim: () => import_colors.dim,
  gray: () => import_colors.gray,
  green: () => import_colors.green,
  greenBright: () => import_colors.greenBright,
  hidden: () => import_colors.hidden,
  inverse: () => import_colors.inverse,
  italic: () => import_colors.italic,
  magenta: () => import_colors.magenta,
  magentaBright: () => import_colors.magentaBright,
  midTrim: () => import_stry.midTrim,
  red: () => import_colors.red,
  redBright: () => import_colors.redBright,
  reset: () => import_colors.reset,
  sleep: () => import_sleep.sleep,
  strTrimMid: () => import_str.strTrimMid,
  strikethrough: () => import_colors.strikethrough,
  stry: () => import_stry.stry,
  stry0: () => import_stry.stry0,
  stryEq: () => import_stry.stryEq,
  underline: () => import_colors.underline,
  white: () => import_colors.white,
  whiteBright: () => import_colors.whiteBright,
  yellow: () => import_colors.yellow,
  yellowBright: () => import_colors.yellowBright
});
module.exports = __toCommonJS(index_exports);
var import_sleep = require("./sleep.js");
var import_stry = require("./stry/stry.js");
var import_str = require("./format/str.js");
var import_date = require("./format/date.js");
var import_Log = require("./Log.js");
var import_colors = require("./colors/index.js");
